Fix deprecations in common module (#22209)

thomasdarimont opened the pull request:
- Use charset in `Encode` class
- Replace reflective call to protected `Liquibase#resetServices()` with call to exposed public method on a custom subclass `KeycloakLiquibase`
- Remove usage of deprecated AccessController class in Reflections
- Mark unused SetAccessibleProvilegedAction and UnsetAccessibleProvilegedAction as deprecated for removal in Keycloak 27

Fixes #22209

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

thomasdarimont committed:
Fix deprecations in common module (#22209)

- Use charset in `Encode` class
- Replace reflective call to protected `Liquibase#resetServices()` with call to exposed public method on a custom subclass `KeycloakLiquibase`
- Remove usage of deprecated AccessController class in Reflections
- Remove unused SetAccessibleProvilegedAction and UnsetAccessibleProvilegedAction

Fixes #22209

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

ahus1 reviewed (changes requested):
Thank you for this PR. I agree the Liquibase parts can be improved by using less reflection. 

The removal of the classes in keycloak-common should be done by first deprecating them, and later removing them in a future issue. 

The change in `Encode.java` should not be a problem as we set `-Dfile.encoding=UTF-8` in the `JAVA_OPTS` in `kc.sh`, still it would be good to mention it in the release notes as it could lead to a different encoding of URL parameters.

thomasdarimont committed:
Mark (Un)SetAccessiblePrivilegedAction as deprecated for removal.

Avoid usage of interfaces used by deprecated AccessController infrastructure.

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont committed:
Add note to release notes about consistent UTF-8 charset usage for URL encoding

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont commented:
@ahus1 thanks for taking a look! I readded the removed classes and marked them for removal in KC27. I also added a brief note to the release notes.

ahus1 commented on common/src/main/java/org/keycloak/common/util/reflections/Reflections.java:
```diff
@@ -988,7 +987,7 @@ public static boolean isPrimitive(Type type) {
      * @throws InstantiationException
      */
-    public static <T> T newInstance(final Class<T> fromClass) throws ClassNotFoundException, IllegalAccessException, InstantiationException {
+    public static <T> T newInstance(final Class<T> fromClass) throws ClassNotFoundException, IllegalAccessException, InstantiationException, InvocationTargetException, NoSuchMethodException {
```
Nitpick: Changing the signature of the method is also a breaking change at least when people want to compile their code. AFAIK this method is not used, so I'd suggest to either keep it as is for now, deprecate it for KC27 and eventually remove it.

ahus1 commented on common/src/main/java/org/keycloak/common/util/reflections/Reflections.java:
```diff
@@ -1006,8 +1005,8 @@ public static <T> T newInstance(final Class<T> fromClass) throws ClassNotFoundEx
      * @throws IllegalAccessException
      * @throws InstantiationException
      */
-    public static <T> T newInstance(final Class<?> type, final String fullQualifiedName) throws ClassNotFoundException, IllegalAccessException, InstantiationException {
```
Nitpick - as above, I suggest deprecation and removal.

ahus1 reviewed (changes requested):
Thank you for the changes. Please add the deprecated methods to `changes-26_0_0.adoc` so people will know that that they'll need to change something.

thomasdarimont committed:
Revised code according to code review

- Added keycloak-common deprecations to changes-26 document
- Avoid breaking API changes

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

ahus1 committed:
Review of the issue

Closes #22209

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 reviewed (dismissed):
Thank you for this PR. I pushed one more commit as the updated class in `common` still contained additional exceptions in its signature. 

Thinking again about the docs, also decided to move all the docs changes to the `upgrading` guide as only those upgrading might be interested in them. The release notes usually only contain the highlights.
